jepsen/src/jepsen/net.clj

```diff
@@ -46,75 +48,118 @@
-          :reliable
-          (throw e))))))
-
-(defn- set-netem!
-  "Sets a network emulation with given `behaviors` to disrupt packets.
-   Shared convenience call for iptables/ipfilter."
-  [net test behaviors]
-  (if (seq behaviors)
-    (do
-      (assert (if (:reorder behaviors)
-                (:delay behaviors)
-                true)
-              "Cannot reorder packets without a delay.")
-      ;; reset node networks to reliable before introducing new queuing discipline
-      (delete-netem! net test)
-      (let [tc-opts (->> behaviors
-                         ;; fill in all unspecified opts with default values
-                         (reduce (fn [acc [behavior opts]]
-                                   (assoc acc behavior (merge (behavior all-packet-behaviors) opts)))
-                                 {})
-                         ;; build a tc cmd line combining all behaviors
-                         (reduce (fn [args [behavior {:keys [time jitter percent correlation distribution] :as _opts}]]
-                                   (case behavior
-                                     :delay
-                                     (concat args [:delay time jitter correlation distribution])
-                                     (:loss :corrupt :duplicate :reorder)
-                                     (concat args [behavior percent correlation])))
-                                 [:qdisc :add :dev :eth0 :root :netem]))]
-        (with-test-nodes test
-          (do
-            (su (exec tc tc-opts))
-            :shaped))))
-    ;; no queuing discipline desired, delete any existing shaping
-    (delete-netem! net test)))
+               :correlation  :75%}
+   :rate      {:rate         :1mbit}})
+
+(defn- behaviors->netem
+  "Given a map of behaviors, returns a sequence of netem options."
+  [behaviors]
+  (->>
+   ; :reorder requires :delay
+   (if (and (:reorder behaviors)
+            (not (:delay behaviors)))
+     (assoc behaviors :delay (:delay all-packet-behaviors))
+     behaviors)
+   ; fill in all unspecified opts with default values
+   (reduce (fn [acc [behavior opts]]
+             (assoc acc behavior (merge (behavior all-packet-behaviors) opts)))
+           {})
+   ; build a tc cmd line combining all behaviors
+   (reduce (fn [args [behavior {:keys [time jitter percent correlation distribution rate] :as _opts}]]
+             (case behavior
+               :delay
+               (concat args [:delay time jitter correlation :distribution distribution])
+               (:loss :corrupt :duplicate :reorder)
+               (concat args [behavior percent correlation])
+               :rate
+               (concat args [:rate rate])))
+           [])))
+
+(defn- net-shape!
+  "Shared convenience call for iptables/ipfilter.
+   Shape the network with tc qdisc, netem, and filter(s) so target nodes have given behavior."
+  [_net {:keys [nodes] :as _test} targets behavior]
+  (let [results (on-many nodes
```

Comment:
BTW, using `(c/on-nodes test nodes (fn [test node] ...))` is a bit faster, since it'll re-use the existing SSH sessions for the test.

Reply:
Noticeably faster, thanks!